File: array/merge_intervals/merge_intervals.js
function mergeIntervals(arr) {
    arr.sort((a, b) => a.start - b.start)
    const newArr = [arr[0]]
    for (const i of arr.slice(1)) {
        const last = newArr.slice(-1)[0]
        if (i.start > last.end) {
            newArr.push(i)
        } else {
            newArr[newArr.length - 1] = {
                start: last.start,
                end: Math.max(i.end, last.end)
            }
        }
    }
    return newArr
}

const given = [[1, 3], [2, 6], [8, 10], [15, 18]]
const givenObj = given.map((interv) => ({
    start: interv[0],
    end: interv[1]
}))
const expected = [[1, 6], [8, 10], [15, 18]]
console.log("input: ", given)
console.log("result: ", mergeIntervals(givenObj).map(i => [i.start, i.end]))
console.log("output should be: ", expected)

File: array/merge_intervals/merge_intervals.py
from collections import namedtuple

Interval = namedtuple('Interval', ['start', 'end'])


# O(n*logn) for sorting. merging takes linear time


def merge_intervals(arr):
    arr.sort(key=lambda x: x.start)
    new = [arr[0]]
    for i in arr[1:]:
        last = new[-1]
        if i.start > last.end:
            new.append(i)
        else:
            new[-1] = Interval(last.start, max(i.end, last.end))
    return new


if __name__ == "__main__":
    given = [[1, 3], [2, 6], [8, 10], [15, 18]]
    given_intervals = [Interval(*i) for i in given]
    expected = [[1, 6], [8, 10], [15, 18]]
    print("input: ", given)
    print("result: ", [[interv.start, interv.end]
                       for interv in merge_intervals(given_intervals)])
    print("output should be: ", expected)
